File: planet-render/src/main/java/au/com/codeka/warworlds/planetrender/TemplatedPerlinNoise.java
package au.com.codeka.warworlds.planetrender;

import au.com.codeka.warworlds.common.PerlinNoise;
import java.util.Random;

/** A {@link PerlinNoise} which takes it's parameters from a {@link Template}. */
public class TemplatedPerlinNoise extends PerlinNoise {
  public TemplatedPerlinNoise(Template.PerlinNoiseTemplate tmpl, Random rand) {
    setRawSeed(rand.nextLong());
    setPersistence(tmpl.getPersistence());
    setStartOctave(tmpl.getStartOctave());
    setEndOctave(tmpl.getEndOctave());

    if (tmpl.getInterpolation() == Template.PerlinNoiseTemplate.Interpolation.None) {
      setInterpolator(new PerlinNoise.NoneInterpolator());
    } else if (tmpl.getInterpolation() == Template.PerlinNoiseTemplate.Interpolation.Linear) {
      setInterpolator(new PerlinNoise.LinearInterpolator());
    } else if (tmpl.getInterpolation() == Template.PerlinNoiseTemplate.Interpolation.Cosine) {
      setInterpolator(new PerlinNoise.CosineInterpolator());
    } else {
      setInterpolator(new PerlinNoise.NoneInterpolator()); // ??
    }
  }
}

File: planet-render/src/main/java/au/com/codeka/warworlds/planetrender/TemplatedPerlinNoise.kt
package au.com.codeka.warworlds.planetrender

import au.com.codeka.warworlds.common.PerlinNoise
import au.com.codeka.warworlds.planetrender.Template.PerlinNoiseTemplate
import au.com.codeka.warworlds.planetrender.Template.PerlinNoiseTemplate.Interpolation
import java.util.*

/** A [PerlinNoise] which takes it's parameters from a [Template].  */
class TemplatedPerlinNoise(tmpl: PerlinNoiseTemplate, rand: Random) : PerlinNoise() {
  init {
    rawSeed = rand.nextLong()
    persistence = tmpl.persistence
    startOctave = tmpl.startOctave
    endOctave = tmpl.endOctave
    interpolator = when (tmpl.interpolation) {
      Interpolation.None -> {
        NoneInterpolator()
      }
      Interpolation.Linear -> {
        LinearInterpolator()
      }
      Interpolation.Cosine -> {
        CosineInterpolator()
      }
      else -> {
        NoneInterpolator() // ??
      }
    }
  }
}
